coreax/metrics.py

```diff
@@ -39,472 +39,114 @@
-    ) -> float:
-        r"""
-        Sum weighted kernel distance between all pairs of points in the two datasets.
-
-        The summation is done in blocks to avoid excessive memory usage.
-
-        :param reference_points: :math:`n \times 1` array
-        :param comparison_points: :math:`m \times 1` array
-        :param reference_weights: :math:`n \times 1` array of weights for
-            ``reference_points``
-        :param comparison_weights: :math:`m \times 1` array of weights for
-            ``comparison_points``
-        :param block_size: Size of matrix blocks to process
-        :return: The sum of pairwise distances between points in ``reference_points``
-            and ``comparison_points``, with contributions weighted as defined by
-            ``reference_weights`` and ``comparison_weights``
-        """
-        # Format inputs
-        reference_points = jnp.atleast_2d(reference_points)
-        comparison_points = jnp.atleast_2d(comparison_points)
-        reference_weights = jnp.atleast_1d(reference_weights)
-        comparison_weights = jnp.atleast_1d(comparison_weights)
-        block_size = max(0, block_size)
-
-        num_reference_points = len(reference_points)
-        num_comparison_points = len(comparison_points)
-
-        # If block_size is larger than both inputs, we don't need to consider block-wise
-        # computation
-        if block_size > max(num_reference_points, num_comparison_points):
-            kernel_weights = (
-                self.kernel.compute(reference_points, comparison_points)
-                * comparison_weights
-            )
-            weighted_pairwise_distance_sum = (
-                reference_weights * kernel_weights.T
-            ).sum()
-
-        else:
-            weighted_pairwise_distance_sum = 0
-
-            try:
-                row_index_range = range(0, num_reference_points, block_size)
-            except ValueError as exception:
-                if block_size == 0:
-                    raise ValueError(
-                        "block_size must be a positive integer"
-                    ) from exception
-                raise
-            except TypeError as exception:
-                raise TypeError("block_size must be a positive integer") from exception
-
-            for i in row_index_range:
-                for j in range(0, num_comparison_points, block_size):
-                    pairwise_distances_part = (
-                        reference_weights[i : i + block_size, None]
-                        * self.kernel.compute(
-                            reference_points[i : i + block_size],
-                            comparison_points[j : j + block_size],
-                        )
-                        * comparison_weights[None, j : j + block_size]
-                    )
-                    weighted_pairwise_distance_sum += pairwise_distances_part.sum()
-
-        return weighted_pairwise_distance_sum
+def _permuted_block_sizes(
```

Comment:
Add some unit tests for this function. It took me a while to work out the mn spaghetti!

Reply:
I've removed this function in favour of `itertools.product` and replaced all the `n` with `x` and all the `m` with `y` as these are much easier to read next to each other.